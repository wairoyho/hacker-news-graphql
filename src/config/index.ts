import convict from "convict";
import convictFormatValidator from "convict-format-with-validator";
import dotenv from "dotenv";
import path from "path";

import configSchema, { ConfigSchema } from "./schema";

dotenv.config();

convict.addFormats(convictFormatValidator);
const config = convict(configSchema);

const stage = config.get("stage");
const stageConfigPath = path.join(__dirname, `config.${stage}`);
const stageConfig = require(stageConfigPath).default;

config.load(stageConfig);

console.log(config.toString());

export type IConfig = convict.Config<ConfigSchema>;

export default config;
