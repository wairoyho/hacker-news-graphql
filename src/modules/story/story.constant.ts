export enum StoryType {
  TOP = "top",
  NEW = "new",
  BEST = "best",
  SHOW = "show",
  JOB = "job",
}
