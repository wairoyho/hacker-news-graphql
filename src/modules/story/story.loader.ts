import DataLoader from "dataloader";

// import CacheService from "@modules/cache/cache.service";

// const ttl = 60 * 5 * 1; // cache for 5 mins
// const cache = new CacheService(ttl);

export type IStoryLoader = DataLoader<any, any, any>;

const storyLoader = (storyRepository, logger): IStoryLoader =>
  new DataLoader(async (keys) => {
    const resp = await storyRepository.getStories(keys);
    logger.debug("rsp", resp);
    return resp;
  });

export default storyLoader;
