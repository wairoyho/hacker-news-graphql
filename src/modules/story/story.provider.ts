import DataLoader from "dataloader";

import StoryLoader from "@modules/story/story.loader";
import StoryRepository from "@modules/story/story.repository";
import StoryService from "@modules/story/story.service";
import StoryResolver from "@modules/story/story.resolvers";

export interface IStoryServices {
  storyRepository?: StoryRepository;
  storyLoader?: DataLoader<any, any, any>;
  storyService?: StoryService;
  storyResolver?: any;
}

enum SERVICES {
  STORY_REPOSITORY = "storyRepository",
  STORY_LOADER = "storyLoader",
  STORY_SERVICE = "storyService",
  STORY_RESOLVER = "storyResolver",
}

const storyProvider = (container, CONTAINER_SERVICE) => {
  container.register(SERVICES.STORY_REPOSITORY, StoryRepository, [
    CONTAINER_SERVICE.API_CLIENT,
    CONTAINER_SERVICE.LOGGER,
  ]);
  container.register(SERVICES.STORY_LOADER, StoryLoader, [
    SERVICES.STORY_REPOSITORY,
    CONTAINER_SERVICE.LOGGER,
  ]);
  container.register(SERVICES.STORY_SERVICE, StoryService, {
    storyRepository: SERVICES.STORY_REPOSITORY,
    storyLoader: SERVICES.STORY_LOADER,
    logger: CONTAINER_SERVICE.LOGGER,
    cache: CONTAINER_SERVICE.CACHE,
  });
  container.register(SERVICES.STORY_RESOLVER, StoryResolver, [
    SERVICES.STORY_SERVICE,
    CONTAINER_SERVICE.LOGGER,
  ]);
};

export default storyProvider;
