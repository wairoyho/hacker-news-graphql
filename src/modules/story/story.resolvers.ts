import { encodeToCursor } from "@src/helpers/cursor";

import { resolveCommentList } from "@modules/comment/comment.resolvers";
import { resolveUser } from "@modules/user/user.resolvers";

const resolveStory = ({ storyService }: IServices) => async (
  parent,
  args,
  ctx,
  info
) => {
  const { id } = args;

  const story = await storyService.getStoryById(id);

  return story;
};

const resolveStoryList = ({ storyService, logger }: IServices) => async (
  parent,
  args,
  ctx
) => {
  logger.info(args);
  const { type, first, last, after, before } = args;

  try {
    const {
      list: stories,
      pageInfo,
      totalCount,
      cursorPrefix,
    } = await storyService.getStories(type, args);

    return {
      edges: stories.map((story) => ({
        cursor: encodeToCursor(story.id, cursorPrefix),
        node: story,
      })),
      nodes: stories,
      pageInfo,
      totalCount: totalCount,
    };
  } catch (e) {
    throw new Error(e);
  }
};

const resolvers = (services: IServices) => ({
  Query: {
    Story: resolveStory(services),
    StoryList: resolveStoryList(services),
  },
  Story: {
    comments: resolveCommentList(services),
    author: resolveUser(services),
  },
});

export default resolvers;
