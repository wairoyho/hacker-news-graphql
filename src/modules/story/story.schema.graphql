type StoryCommentsConnection {
  pageInfo: PageInfo!
  edges: [StoryCommentsEdge]
  nodes: [Comment]
  totalCount: Int
}

type StoryCommentsEdge {
  cursor: String!
  node: Comment
}

enum StoryType {
  # TOP
  # NEW
  # BEST
  # SHOW
  # JOB
  story
  comment
  job
  poll
  pollopt
}

type Story implements Node {
  author: User
  by: String
  descendants: Int
  id: ID!
  """
  the list of Comments in this Story
  """
  kids: [Int]
  comments(
    """
    Returns the elements in the list that come after the specified cursor.
    """
    after: String
    """
    Returns the elements in the list that come before the specified cursor.
    """
    before: String
    """
    Returns the first n elements from the list.
    """
    first: Int
    """
    Returns the last n elements from the list.
    """
    last: Int
  ): StoryCommentsConnection
  score: Int
  time: Int
  title: String
  type: StoryType
  url: String
}

enum StoryTypeInput {
  TOP
  NEW
  BEST
  SHOW
  JOB
}

type StoryListConnection {
  pageInfo: PageInfo!
  edges: [StoryEdge]
  nodes: [Story]
  totalCount: Int
}

type StoryEdge {
  cursor: String!
  node: Story
}

type Query {
  Story(id: ID): Story
  StoryList(
    type: StoryTypeInput
    first: Int
    after: String
    last: Int
    before: String
  ): StoryListConnection
}
