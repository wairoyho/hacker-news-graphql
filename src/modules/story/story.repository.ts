import { StoryType } from "./story.constant";
import { Story } from "./story.entity";

class StoryRepository {
  apiClient;
  constructor(apiClient) {
    this.apiClient = apiClient;
  }

  async getItem(itemId: string): Promise<Story> {
    const res = await this.apiClient.get(`/item/${itemId}.json`);
    return res.data;
  }

  async getStory(storyId: string) {
    return this.getItem(storyId);
  }

  async getStories(storyIds: any[]) {
    const result = await Promise.all(
      storyIds.map(async (storyId) => {
        const story = await this.getStory(storyId);
        return story;
      })
    );

    return storyIds.map(
      (key) =>
        result.find(({ id }) => key == id) || new Error(`No result for ${key}`)
    );
  }

  async getStoryList(type: StoryType): Promise<number[]> {
    const res = await this.apiClient.get(`/${type.toLowerCase()}stories.json`);
    return res.data;
  }
}

export default StoryRepository;
