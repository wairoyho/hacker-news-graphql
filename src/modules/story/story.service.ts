import DataLoader from "dataloader";

import {
  encodeToCursor,
  destructureCursor,
  getSelectedList,
  IPaginationArguments,
} from "@src/helpers/cursor";
import { ILogger } from "@modules/logger/logger.service";
import Cache from "@modules/cache/cache.service";

import { StoryType } from "./story.constant";
import StoryRepository from "./story.repository";

const extractCursorPrefix = (cursor) => {
  return destructureCursor(cursor).slice(0, -1);
};

const getCursorPrefix = (type, before = undefined, after = undefined) => {
  let prefix = "";
  if (before || after) {
    const snapshotVersion = before
      ? extractCursorPrefix(before)
      : extractCursorPrefix(after);
    prefix = snapshotVersion.join(":");
  } else {
    prefix = createStoryListCursorPrefix(type);
  }
  return prefix;
};

const createStoryListCursorPrefix = (type) => {
  // pre min snapshoot
  const coeff = 1000 * 10 * 1;
  const snapshotId = new Date(Math.floor(Date.now() / coeff) * coeff).getTime();

  return `story_list:${type.toLowerCase()}:${snapshotId}`;
};

class StoryService {
  storyRepository: StoryRepository;
  storyLoader: DataLoader<any, any, any>;
  logger: ILogger;
  cache: Cache;

  constructor({ storyRepository, storyLoader, logger, cache }) {
    this.storyRepository = storyRepository;
    this.storyLoader = storyLoader;
    this.logger = logger;
    this.cache = cache;
  }

  async getStoryById(storyId: string) {
    return await this.storyLoader.load(storyId);
  }

  async getStories(type: StoryType, args: IPaginationArguments) {
    const { first, last, before, after } = args;

    let snapshotPrefix = getCursorPrefix(type, before, after);

    let storyIdList = [];
    // if (before || after) {
    //   const snapshotVersion = before
    //     ? extractCursorPrefix(before)
    //     : extractCursorPrefix(after);

    //   snapshotPrefix = snapshotVersion.join(":");

    //   if (this.cache.has(snapshotVersion)) {
    //     storyIdList = (await this.cache.get(snapshotPrefix)) ?? [];
    //   }
    // } else {
    //   storyIdList = await this.cache.get(snapshotPrefix, () =>
    //     this.storyRepository.getStoryList(type)
    //   );
    // }
    console.log("snapshotPrefix", snapshotPrefix);
    if (before || after) {
      storyIdList = (await this.cache.get(snapshotPrefix)) ?? [];
    } else {
      storyIdList = await this.cache.get(snapshotPrefix, () =>
        this.storyRepository.getStoryList(type)
      );
    }

    let { list, pageInfo } = getSelectedList(storyIdList, args);
    const stories = await this.storyLoader.loadMany(list);

    const filteredStories = stories.filter((story) => story !== null);

    if (filteredStories.length > 0) {
      this.logger.debug("stories", stories);
      pageInfo.startCursor = encodeToCursor(stories[0].id, snapshotPrefix);
      pageInfo.endCursor = encodeToCursor(
        stories[stories.length - 1].id,
        snapshotPrefix
      );
    }

    return {
      list: filteredStories,
      pageInfo,
      totalCount: storyIdList.length,
      cursorPrefix: snapshotPrefix,
    };
  }

  async getLastestStories(args: IPaginationArguments) {
    const { first, last, before, after } = args;

    const storyIdList = await this.storyRepository.getStoryList(StoryType.NEW);

    let { list, pageInfo } = getSelectedList(storyIdList, args);
    const stories = await this.storyLoader.loadMany(list);

    const filteredStories = stories.filter((story) => story !== null);

    if (filteredStories.length > 0) {
      pageInfo.startCursor = encodeToCursor(
        stories[0].id,
        createStoryListCursorPrefix(StoryType.NEW)
      );
      pageInfo.endCursor = encodeToCursor(
        stories[stories.length - 1].id,
        createStoryListCursorPrefix(StoryType.NEW)
      );
    }

    return { list: filteredStories, pageInfo, totalCount: storyIdList.length };
  }
}

export default StoryService;
