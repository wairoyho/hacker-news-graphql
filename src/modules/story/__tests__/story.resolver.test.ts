// @ts-nocheck

import config from "@config";
import ApiClient from "@src/helpers/apiClient";
import Logger from "@modules/logger/logger.service";

import { StoryType } from "../story.constant";
import StoryLoader from "../story.loader";
import StoryRepository from "../story.repository";
import StoryService from "../story.service";
import storyResovlers from "../story.resolvers";

import storyApiResonse from "../__fixtures__/storyApiResponse";

jest.mock("@src/config");
jest.mock("@src/helpers/apiClient");
jest.mock("@modules/logger/logger.service");
jest.mock("../story.loader");
jest.mock("../story.repository");
// jest.mock("../story.service");

describe("Story resolver", () => {
  const logger = Logger({ config });
  const apiClient = ApiClient(config, logger);
  const storyRepo = new StoryRepository(apiClient);
  const storyLoader = StoryLoader(storyRepo, logger);
  const storyService = new StoryService(storyRepo, storyLoader, logger);

  it("pass thought", () => {
    expect(null).toEqual(null);
  });

  it("will fail every time", () => {
    const user = {
      createdAt: new Date("2020-12-02T09:23:54.907Z"),
      id: 14,
      name: "LeBron James",
    };

    expect(user).toMatchSnapshot();
  });

  // it("query: Story", async () => {
  //   const storyId = "10001";
  //   const resolveArgs = {
  //     parent: undefined,
  //     args: {
  //       id: storyId,
  //     },
  //   };
  //   const expectedResult = storyApiResonse.getById(storyId).data;

  //   const resovlers = storyResovlers({ storyService, logger });
  //   const result = await resovlers.Query.Story(
  //     resolveArgs.parent,
  //     resolveArgs.args
  //   );

  //   expect(result).toBe(expectedResult);
  // });

  // it("query: StoryList", async () => {
  //   const storyIds = ["10001", "10002", "10003"];
  //   const resolveArgs = {
  //     parent: undefined,
  //     args: {
  //       type: "new",
  //       first: 3,
  //     },
  //   };

  //   const mockDataList = storyIds.map((storyId) =>
  //     storyApiResonse.getById(storyId)
  //   );

  //   const expectedResult = {
  //     list: mockDataList.map((mockData) => mockData.data),
  //     pageInfo: {
  //       hasNextPage: false,
  //       hasPreviousPage: false,
  //       startCursor: "MTAwMDE=",
  //       endCursor: "MTAwMDM=",
  //     },
  //     totalCount: resolveArgs.args.first,
  //   };

  //   const resovlers = storyResovlers({ storyService, logger });
  //   const result = await resovlers.Query.StoryList(
  //     resolveArgs.parent,
  //     resolveArgs.args
  //   );

  //   console.log("result", JSON.stringify(result, null, 2));
  //   expect(result).toBe(expectedResult);
  // });
});
