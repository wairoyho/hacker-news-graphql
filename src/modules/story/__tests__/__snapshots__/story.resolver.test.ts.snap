// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Story resolver will fail every time 1`] = `
Object {
  "createdAt": 2020-12-02T09:23:54.907Z,
  "id": 14,
  "name": "LeBron James",
}
`;
